### DISCUSSION TITLE
#3343

### DISCUSSION THREAD

**Discussion by wuliaodexiaoluo** [Meloong-Git/PCL#3611] [讨论]
#3343 版聊专区

**Comment by wuliaodexiaoluo** [comment]
> 什么逆天bug。。。 让龙猫做个名字是EW+皮肤是Steve时拒绝启动或警告吧？顺便去mojang交个bug

不知这个
我的 GitHub 用户名也不能作为启动用户名。。。 
~这什么逆天 Bug 。。。。。~
![image](https://github.com/Hex-Dragon/PCL2/assets/160083863/42ee8d6c-0fe3-46ad-af2b-12d2986c0b65)
这个问题。。。就极度缺德。。。对。。没错。。。。就挺缺德的。。。就

**Comment by mc-124** [comment]
证实与MC无关，是PCL2导致的
但是是哪导致的，为什么导致的，也不知道，要看生成uuid的逻辑那(ModLaunch.vb)但我vb不太好也看不太懂

**Reply by wuliaodexiaoluo** [reply]
我还在发愁到底哪部分是负责启动的，我去看看

**Reply by mc-124** [reply]
换成alex后消失的俩0回来了

**Reply by mc-124** [reply]
@wuliaodexiaoluo 你去line 984附近看看，那是根据不同皮肤生成不同uuid的部分

**Reply by mc-124** [reply]
明明代码几乎一模一样……但为什么会整出这逆天bug

**Reply by mc-124** [reply]
看了一会发现问题很可能在ModBase.vb:1524和ModLaunch.vb:1046 应该是根据字符串hash生成uuid导致的问题，PCL2离线登录使用用户名的hash去生成uuid

**Reply by wuliaodexiaoluo** [reply]
我去看一眼

**Reply by wuliaodexiaoluo** [reply]
ModBase 那段 MD5 应该是没问题的，不太可能出现小于 128 bit 的 Md5
1046 那一段应该也没问题，不然 Substring 会抛出异常
12 + 1 + 3 + 1 + 15 = 32
看起来完全没问题的说

**Reply by wuliaodexiaoluo** [reply]
看了半天我觉得最可能的是这段
```
Private Function McLoginLegacyUuidWithCustomSkin(UserName As String, SkinType As Integer, SkinName As String) As String
        Dim Uuid As String = McLoginLegacyUuid(UserName)
        '根据离线皮肤获取实际使用的 Uuid
        Select Case SkinType
            Case 0
                '默认，不需要处理
            Case 1
                'Steve
                Do Until McSkinSex(Uuid) = "Steve"
                    If Uuid.EndsWithF("FFFFF") Then Uuid = Mid(Uuid, 1, 32 - 5) & "00000"
                    Uuid = Mid(Uuid, 1, 32 - 5) & (Long.Parse(Right(Uuid, 5), Globalization.NumberStyles.AllowHexSpecifier) + 1).ToString("X")
                Loop
            Case 2
                'Alex
                Do Until McSkinSex(Uuid) = "Alex"
                    If Uuid.EndsWithF("FFFFF") Then Uuid = Mid(Uuid, 1, 32 - 5) & "00000"
                    Uuid = Mid(Uuid, 1, 32 - 5) & (Long.Parse(Right(Uuid, 5), Globalization.NumberStyles.AllowHexSpecifier) + 1).ToString("X")
                Loop
```
Mid 没满没有异常抛出，如果传入小于32位就把全部写进去，然后 Replace 一下可能就是小于 32 位的UUID
至于可能错误修改 UUID 的方法我没找到，估计是只能在这里做判断，不满就填充到 32 位
